### PR TITLE
GSoC 2024 - Streaming H5MD trajectories from cloud services

### PR DESCRIPTION
**Pull request by ljwoods2**
Added blog post + updated link

### CODE REVIEW

File: _posts/2024-08-03-gsoc2024_woods.markdown [hunk 1]
```diff
@@ -0,0 +1,261 @@
+---
+layout: single
+title:  "GSoC 2024 - Streaming H5MD trajectories from the cloud"
+categories: zarrtraj
+---
+
+## What is Zarrtraj?
+
+[Zarrtraj](https://github.com/Becksteinlab/zarrtraj) is an [MDAKit](https://mdakits.mdanalysis.org/about.html) for storing and analyzing trajectories in [MDAnalysis](https://github.com/MDAnalysis/mdanalysis) from the cloud, representing a major milestone towards the MDAnalysis 3.0 goal of cloud streaming as well as a proof-of-concept for a new paradigm in the field of molecular dynamics as a whole.
```
Comment (jennaswa) [inline _posts/2024-08-03-gsoc2024_woods.markdown]:
```suggestion
[Zarrtraj](https://github.com/Becksteinlab/zarrtraj) is an [MDAKit](https://mdakits.mdanalysis.org/about.html) for storing and analyzing trajectories in [MDAnalysis](https://github.com/MDAnalysis/mdanalysis) from the cloud, representing a major milestone towards the [MDAnalysis 3.0 goal of cloud streaming](https://www.mdanalysis.org/2023/10/25/towards_3.0) as well as a proof-of-concept for a new paradigm in the field of molecular dynamics as a whole.
```

Comment (jennaswa) [reply _posts/2024-08-03-gsoc2024_woods.markdown]: Might be good to link the 3.0 roadmap for this.

File: _posts/2024-08-03-gsoc2024_woods.markdown [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+It can stream [H5MD](https://www.nongnu.org/h5md/h5md.html) trajectories from cloud storage 
+providers including AWS S3, Google Cloud buckets, and Azure Blob storage and data lakes.
```
Comment (jennaswa) [inline _posts/2024-08-03-gsoc2024_woods.markdown]:
```suggestion
providers including [AWS S3](https://aws.amazon.com/s3/), [Google Cloud](https://cloud.google.com/) buckets, and [Azure Blob](https://azure.microsoft.com/en-us/products/storage/blobs) storage and data lakes.
```

Comment (jennaswa) [reply _posts/2024-08-03-gsoc2024_woods.markdown]: More links

File: _posts/2024-08-03-gsoc2024_woods.markdown [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+Using Zarrtraj, anyone can reproduce your analyses or train their ML model on trajectory data
```
Comment (jennaswa) [inline _posts/2024-08-03-gsoc2024_woods.markdown]:
```suggestion
Using Zarrtraj, anyone can reproduce your analyses or train their machine learning (ML) model on trajectory data
```

File: _posts/2024-08-03-gsoc2024_woods.markdown [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+without ever having to download massive files to their disk.
+
+This is possible thanks to [Zarr](https://github.com/zarr-developers/zarr-python), [fsspec](https://github.com/fsspec/filesystem_spec),
+and [kerchunk](https://github.com/fsspec/kerchunk), packages that have created a foundation for storing and interacting with
```
Comment (jennaswa) [inline _posts/2024-08-03-gsoc2024_woods.markdown]:
```suggestion
and [kerchunk](https://github.com/fsspec/kerchunk) packages that have created a foundation for storing and interacting with
```

File: _posts/2024-08-03-gsoc2024_woods.markdown (hunk 1 above)
Comment (jennaswa) [inline _posts/2024-08-03-gsoc2024_woods.markdown]:
```suggestion
[Zarrtraj](https://github.com/Becksteinlab/zarrtraj) is an [MDAKit](https://mdakits.mdanalysis.org/about.html) for storing and analyzing trajectories in [MDAnalysis](https://github.com/MDAnalysis/mdanalysis) from the cloud, representing a major milestone towards the MDAnalysis 3.0 goal of cloud streaming as well as a proof-of-concept for a new paradigm in the field of molecular dynamics (MD) as a whole.
```

Comment (jennaswa) [reply _posts/2024-08-03-gsoc2024_woods.markdown]: Define for use later in the blog (line 31).

File: _posts/2024-08-03-gsoc2024_woods.markdown [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+In this project, we also decided to experiment with storing trajectories directly in Zarr-backed files
+using the same specification that H5MD uses, so Zarrtraj can read both `.h5md` and H5MD-formatted `.zarrmd`
+files. See [this explanation of the modified format](https://zarrtraj.readthedocs.io/en/latest/zarrmd-file-spec/v0.2.0.html)
+to learn more.
+
+While this GSoC project started with the goal of building a new, Zarr-backed trajectory format, we pivoted
+to making the existing H5MD format streamable after getting feedback
+from the community that supporting widely adopted formats in the MD ecosystem makes code more sustainable and simplifies tool adoption.
+
+The next section is a walkthrough of Zarrtraj's features and usage also available
```
Comment (jennaswa) [inline _posts/2024-08-03-gsoc2024_woods.markdown]:
```suggestion
The next section is a walkthrough of Zarrtraj's features and usage (also available
```

File: _posts/2024-08-03-gsoc2024_woods.markdown [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+[here](https://zarrtraj.readthedocs.io/en/latest/walkthrough.html).
```
Comment (jennaswa) [inline _posts/2024-08-03-gsoc2024_woods.markdown]:
```suggestion
[here](https://zarrtraj.readthedocs.io/en/latest/walkthrough.html)).
```

File: _posts/2024-08-03-gsoc2024_woods.markdown [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+### Reading H5MD trajectories from cloud storage
+
+#### Uploading your H5MD file
+
+First, upload your H5MD trajectories to an AWS S3 bucket. This requires that an S3 Bucket is setup and configured for 
+write access using the credentials stored in "sample_profile". If you've never configured an S3 Bucket before, see
+[this guide](https://docs.aws.amazon.com/AmazonS3/latest/userguide/creating-bucket.html). You can setup a profile to easily manage AWS
+credentials using [this VSCode extension](https://marketplace.visualstudio.com/items?itemName=AmazonWebServices.aws-toolkit-vscode).
+Here is a sample profile (stored in ~/.aws/credentials) where 
+[the key is an access key associated with a user that has read and write permissions for the bucket](https://stackoverflow.com/questions/50802319/create-a-single-iam-user-to-access-only-specific-s3-bucket)
```
Comment (jennaswa) [inline _posts/2024-08-03-gsoc2024_woods.markdown]:
```suggestion
[the key is an access key associated with a user that has read and write permissions for the bucket](https://stackoverflow.com/questions/50802319/create-a-single-iam-user-to-access-only-specific-s3-bucket).
```

File: _posts/2024-08-03-gsoc2024_woods.markdown [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+                precision=3,
+                compressor=numcodecs.Blosc(cname="zstd", clevel=9)) as W:
+                for ts in u.trajectory:
+                    W.write(u.atoms)
+
+```
+
+If you have additional questions, please don't hesitate to open a discussion on the [zarrtarj github](https://github.com/Becksteinlab/zarrtraj).
+The [MDAnalysis discord](https://discord.com/channels/807348386012987462/) is also a 
+great resource for asking questions and getting involved in MDAnalysis.
```
Comment (jennaswa) [inline _posts/2024-08-03-gsoc2024_woods.markdown]:
```suggestion
great resource for asking questions and getting involved in MDAnalysis; instructions for [joining the MDAnalysis discord server](https://www.mdanalysis.org/#participating) can be found on the MDAnalysis website.
```

File: _posts/2024-08-03-gsoc2024_woods.markdown [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+
+### Asynchronous Reading
+
+The performance impact of network IO could be reduced by creating a multi-threaded `ZARRH5MDReader` that isn't blocked
+by analysis code executing. The reader could eagerly load the cache with the next frames
+the analysis code will need to reduce the impact of network IO on exeution time.
+
+## Acknowledgements
+
+Thank you to Dr. Hugo MacDermott-Opeskin and Dr. Yuxuan Zhuang for their mentorship and feedback
```
Comment (jennaswa) [inline _posts/2024-08-03-gsoc2024_woods.markdown]:
```suggestion
Thank you to Dr. Hugo MacDermott-Opeskin ([@hmacdope](https://github.com/hmacdope)) and Dr. Yuxuan Zhuang ([@yuxuanzhuang](https://github.com/yuxuanzhuang)) for their mentorship and feedback
```

File: _posts/2024-08-03-gsoc2024_woods.markdown [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+throughout this project and to Dr. Jenna Swarthout Goddard for supporting the GSoC program
```
Comment (jennaswa) [inline _posts/2024-08-03-gsoc2024_woods.markdown]:
```suggestion
throughout this project and to Dr. Jenna Swarthout Goddard ([@jennaswa](https://github.com/jennaswa)) for supporting the GSoC program 
```

Comment (jennaswa) [inline _posts/2024-08-03-gsoc2024_woods.markdown]:
```suggestion
throughout this project and to Dr. Jenna Swarthout Goddard for supporting the [Google Summer of Code (GSoC) program 
```

File: _posts/2024-08-03-gsoc2024_woods.markdown [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+
+The performance impact of network IO could be reduced by creating a multi-threaded `ZARRH5MDReader` that isn't blocked
+by analysis code executing. The reader could eagerly load the cache with the next frames
+the analysis code will need to reduce the impact of network IO on exeution time.
+
+## Acknowledgements
+
+Thank you to Dr. Hugo MacDermott-Opeskin and Dr. Yuxuan Zhuang for their mentorship and feedback
+throughout this project and to Dr. Jenna Swarthout Goddard for supporting the GSoC program 
+at MDAnalysis.
```
Comment (jennaswa) [inline _posts/2024-08-03-gsoc2024_woods.markdown]:
```suggestion
at MDAnalysis](https://summerofcode.withgoogle.com/organizations/mdanalysis/programs/2024/projects).
```

File: _posts/2024-08-03-gsoc2024_woods.markdown [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+
+I also want to thank Dr. Oliver Beckstein and Edis Jakupovic for lending their expertise
```
Comment (jennaswa) [inline _posts/2024-08-03-gsoc2024_woods.markdown]:
```suggestion
I also want to thank Dr. Oliver Beckstein ([@orbeckst](https://github.com/orbeckst)) and Edis Jakupovic ([@edisj](https://github.com/edisj) for lending their expertise
```

File: _posts/2024-08-03-gsoc2024_woods.markdown [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+in H5MD and all things MDAnalysis.
+
+Finally, another thanks to Martin Durant, author of Kerchunk, who was incredibly helpful in refining and merging
```
Comment (jennaswa) [inline _posts/2024-08-03-gsoc2024_woods.markdown]:
```suggestion
Finally, another thanks to Martin Durant ([@martindurant](https://github.com/martindurant)), author of Kerchunk, who was incredibly helpful in refining and merging
```

File: _posts/2024-08-03-gsoc2024_woods.markdown [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+- <https://github.com/fsspec/kerchunk/pull/463>
+
+## Extras: Lessons Learned
+
+Here are a bunch of random things I learned while doing this project!
+
+- Any time you need to run code that will take several hours to execute
+  due to file size or some other factor, create a minimal, quickly-executing 
+  example of the code to work out bugs before running the full thing. You will
+  save yourself so, so much frustration
```
Comment (jennaswa) [inline _posts/2024-08-03-gsoc2024_woods.markdown]:
```suggestion
  save yourself so, so much frustration.
```

File: _posts/2024-08-03-gsoc2024_woods.markdown [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+across the trajectory. Early benchmarks show that analysis based on such a lazy array
+can outperform `Timestep`-based analysis.
+
+### Asynchronous Reading
+
+The performance impact of network IO could be reduced by creating a multi-threaded `ZARRH5MDReader` that isn't blocked
+by analysis code executing. The reader could eagerly load the cache with the next frames
+the analysis code will need to reduce the impact of network IO on exeution time.
+
+## Acknowledgements
```
Comment (jennaswa) [inline _posts/2024-08-03-gsoc2024_woods.markdown:185]: It might be a good idea to thank Google here as well :)

**Review by jennaswa — APPROVED** [review]
LGTM - thanks, @ljwoods2 ! I left only a few small suggestions for adding relevant links and a couple of grammatical fixes.

I'm also going to copy your mentors (@hmacdope and @yuxuanzhuang) and honorary mentor (@orbeckst) into this PR to see if they have any feedback.

**Review by hmacdope — APPROVED** [review]
LGTM

**Comment by orbeckst** [comment]
Sorry, don’t have time to review. I trust the mentors!Am 8/15/24 um 11:16 schrieb Hugo MacDermott-Opeskin ***@***.***>:﻿
@hmacdope approved this pull request.

LGTM

File: _posts/2024-08-03-gsoc2024_woods.markdown (hunk 15 above)
Comment (ljwoods2) [reply _posts/2024-08-03-gsoc2024_woods.markdown:185]: Just added it!

**Pull request merged**